src/pe/clr.rs

```diff
@@ -0,0 +1,658 @@
+            .field("streams", &self.streams)
+            .finish()
+    }
+}
+
+/// Represents an individual stream within the .NET storage section, which may contain
+/// metadata or other data specific to the .NET assembly structure.
+#[derive(PartialEq, Copy, Clone, Default)]
+pub struct StorageStream<'a> {
+    /// Indicates the offset, in bytes, from the beginning of the file to this stream.
+    pub offset: u32,
+    /// Indicates the size of this stream, in bytes.
+    pub size: u32,
+    /// Indicates the name of the stream as a null-terminated ANSI string.
+    /// This field can hold up to 32-bytes long, including the null-terminator.
+    pub name: &'a [u8],
+}
+
+impl fmt::Debug for StorageStream<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StorageStream")
+            .field("offset", &format_args!("{:#x}", &self.offset))
+            .field("size", &format_args!("{:#x}", &self.size))
+            .field("name", &self.to_name_string())
+            .field("name_slice", &format_args!("{:02x?}", self.name))
+            .finish()
+    }
+}
+
+impl<'a> StorageStream<'a> {
+    pub fn parse(bytes: &'a [u8], offset: &mut usize) -> error::Result<Self> {
+        let offset_ = bytes.gread_with::<u32>(offset, scroll::LE)?;
+        let size = bytes.gread_with::<u32>(offset, scroll::LE)?;
+        let name_size =
+            &bytes[*offset..].iter().take_while(|x| *x != &0).count() * core::mem::size_of::<u8>();
+        if *offset + name_size as usize + core::mem::size_of::<u8>() > bytes.len() {
+            return Err(error::Error::Malformed(format!(
+                "StorageStream name offset ({:#x}) and size ({:#x}) exceeds bytes slice ({:#x})",
+                *offset,
+                name_size,
+                bytes.len()
+            )));
+        }
+        let name = &bytes[*offset..*offset + name_size + core::mem::size_of::<u8>()];
+        *offset += utils::align_up(name.len(), core::mem::size_of::<u32>());
+        Ok(Self {
+            offset: offset_,
+            size,
+            name,
+        })
+    }
+
+    /// Converts [`StorageStream::name`] slice into a implicit [`Cow`]
+    pub fn to_name_string(&self) -> Cow<'a, str> {
+        // Strip null-terminator if any
+        self.name
+            .strip_suffix(&[0u8])
+            .map(String::from_utf8_lossy)
+            .unwrap_or_else(|| String::from_utf8_lossy(&self.name))
+    }
+}
+
+/// Iterator over [`StorageStream`] (a CLR section) entries in [`Cor20Header::metadata`].
+#[derive(Debug, Copy, Clone)]
+pub struct ClrSectionIterator<'a> {
+    /// The storage header, used to acquire the number of CLR sections
+    pub storage_header: StorageHeader,
+    /// The raw data that scoped to the appropriate offset at the end of [`StorageHeader`]
+    pub data: &'a [u8],
+    /// Internal counter since there are no way to know the size of [`ClrSectionIterator::data`] at the ctor.
+    pub(self) index_cursor: usize,
```

Comment:
hah i've never seen `pub(self)`; how is this different than no visibility annotation at all (e.g., private)?